state/application.go

```diff
@@ -2612,11 +2612,17 @@ func (a *Application) removeUnitOps(u *Unit, asserts bson.D, op *ForcedOperation
 	if u.doc.CharmURL != nil {
 		// If the unit has a different URL to the application, allow any final
 		// cleanup to happen; otherwise we just do it when the app itself is removed.
-		maybeDoFinal := u.doc.CharmURL != a.doc.CharmURL
+		maybeDoFinal := *u.doc.CharmURL != a.doc.CharmURL.String()
```

Comment:
```suggestion
		maybeDoFinal := u.doc.CharmURL != nil && *u.doc.CharmURL != a.doc.CharmURL.String()
```

Reply:
This is inside a block that already checks that.

Reply:
doh! sorry, didn't spot that.